compiler/psi/src/org/jetbrains/kotlin/parsing/KotlinExpressionParsing.java

```diff
@@ -1493,6 +1493,11 @@ private void parseFor() {
                 error("Expecting a variable name");
             }
 
+            TokenSet recoverySet = TokenSet.create(RPAR, LBRACE, RBRACE, EOL_OR_SEMICOLON, EOF);
```

Comment:
What about cases with chars from recovery set like `for (i in 1..3)) {break;}`? The strange semantic errors will still be reported there. I think we should report an error not only on chars not from the recovery set.

Reply:
I don't fully understand what this means. It seems that it's unclear in general when to parse char from recovery set as _"unexpected tokens in 'for' loop"_. For example the case `for (i in 1..2) foo())`: foo() looks more like loop body with extra ')' and regard ') foo()' as error in 'for' declaration can be unexpected. And here` { for (i in 1..2 }` the try to parse '}' inside 'for' will break all parsing further. So should we handle  only case when there is char from recovery set and ')' right after it, e.g. `for (i in 1..2)) {}` or `for (i in 1..2;) {}`?